### DISCUSSION TITLE
Collection - Get specific keys in multidimensional arrays

### DISCUSSION THREAD

**Discussion by ecrmnn** [laravel/framework#32991] [General]
I keep coming in situations where I want to filter out specific keys in a collection based on a multidimensional array.

I've previously made a [PR](https://github.com/laravel/framework/pull/29055), but it was a breaking change when extending the functionality of the `only` method. I'm sure this is a great addition to Collection, but I want to discuss the naming of the method.

---

#### Alternative 1: Accept passing an array to `filter`

In [Pandas](https://pandas.pydata.org/pandas-docs/stable/index.html) you can pass an array to the `filter` method and it will filter out the specific columns. I don't think this would be a breaking change, but would extend the `filter` method.

```php
$people = collect([
    [
        'first' => 'John',
        'last' => 'Doe',
        'name' => 'John Doe',
        'email' => 'john.doe@gmail.com',
    ],
    [
        'first' => 'Jane',
        'last' => 'Doe',
        'name' => 'Jane Doe',
        'email' => 'jane.doe@gmail.com',
    ],
]);

$people->filter(['first', 'last', 'email'])->all();

//=> [
//=>     [
//=>         'first' => 'John',
//=>         'last' => 'Doe',
//=>         'email' => 'john.doe@gmail.com',
//=>     ],
//=>     [
//=>         'first' => 'Jane',
//=>         'last' => 'Doe',
//=>         'email' => 'jane.doe@gmail.com',
//=>     ],
//=> ]
```

#### Alternative 2: Accept passing an array to `map`

When you have to do this manually you would use `map` to return a new collection with the keys you want. Passing an array to map wouldn't be a breaking change, but extend the functionality.

#### Alternative 3: Add `columns` method

Follow the naming of the [`array_column`](https://www.php.net/manual/en/function.array-column.php) method. I would suggest to use `columns` instead of `column`

```php
$people->columns(['first', 'last', 'email'])->all();
```

#### Alternative 4: Add `select` method

Follow the naming of Eloquent when selecting columns.

```php
$people->select(['first', 'last', 'email'])->all();

```

#### Alternative 5: Add `attributes` method

```php
$people->attributes(['first', 'last', 'email'])->all();

```

#### Alternative 6: Add `properties` method

```php
$people->properties(['first', 'last', 'email'])->all();

```

----

### So what do you think? 💁‍♂️
- `filter`
- `columns`
- `select`
- `attributes`
- `properties`

----- 

Update: Added `map` as an alternative
Update: Fixed typo in code example

**Comment by knutle** [comment]
Personally I think select feels most appropriate for this!

**Comment by fsrocha-dev** [comment]
Certainly Filter gets better.

**Comment by smajohusic** [comment]
Attributes feels more natural

**Comment by zotopteam** [comment]
we can use this method:

$people->map->only(['first', 'last', 'email']);

**Comment by larskoole** [comment]
The discussion is pretty old so I don't know if it's fixed already but I'm using the following code (added it to the boot method in the AppServiceProvider.php) to mimic alternative 3:

```PHP
Collection::macro('columns', function ($columns) {
    if (!is_array($columns)) {
        $columns = [$columns];
    }

    return $this->map(function ($item) use ($columns) {
        $newItem = [];
        foreach ($columns as $column) {
            if (array_key_exists($column, $item)) {
                $newItem[$column] = $item[$column];
            }
        }

        return $newItem;
    });
});
```
